lxd/storage/block/discard.go

```diff
@@ -0,0 +1,264 @@
+		}
+
+		return nil
+	}
+
+	// Blocks are trickier to reset with options varying based on disk features.
+	// We use a set of 3 markers to validate whether it was reset.
+	marker := []byte("LXD")
+	markerLength := int64(len(marker))
+	markerOffsetStart := blockOffset
+	markerOffsetMiddle := blockOffset + ((size - blockOffset) / 2)
+	markerOffsetEnd := size - markerLength
+
+	if markerOffsetStart+markerLength > size {
+		// No markers can fit.
+		markerOffsetStart = -1
+		markerOffsetMiddle = -1
+		markerOffsetEnd = -1
+	} else {
+		if markerOffsetMiddle <= markerOffsetStart+markerLength {
+			// Middle marker goes over start marker.
+			markerOffsetMiddle = -1
+		}
+
+		if markerOffsetEnd <= markerOffsetMiddle+markerLength {
+			// End marker goes over middle marker.
+			markerOffsetEnd = -1
+		}
+
+		if markerOffsetEnd <= markerOffsetStart+markerLength {
+			// End marker goes over start marker.
+			markerOffsetEnd = -1
+		}
+	}
+
+	writeMarkers := func(fd *os.File) error {
+		for _, offset := range []int64{markerOffsetStart, markerOffsetMiddle, markerOffsetEnd} {
+			if offset < 0 {
+				continue
+			}
+
+			// Write the marker at the set offset.
+			n, err := fd.WriteAt(marker, offset)
+			if err != nil {
+				return err
+			}
+
+			if n != int(markerLength) {
+				return fmt.Errorf("Only managed to write %d bytes out of %d of the %d offset marker", n, markerLength, offset)
+			}
+		}
+
+		return nil
+	}
+
+	checkMarkers := func(fd *os.File) (int, error) {
+		found := 0
+
+		for _, offset := range []int64{markerOffsetStart, markerOffsetMiddle, markerOffsetEnd} {
+			if offset < 0 {
+				found++
+				continue
+			}
+
+			buf := make([]byte, markerLength)
+
+			// Read the marker from the offset.
+			n, err := fd.ReadAt(buf, offset)
+			if err != nil {
+				return found, err
+			}
+
+			if n != int(markerLength) {
+				return found, fmt.Errorf("Only managed to read %d bytes out of %d of the %d offset marker", n, markerLength, offset)
+			}
+
+			// Check if we found it.
+			if string(buf) == string(marker) {
+				found++
+			}
+		}
+
+		return found, nil
+	}
+
+	// Write and check an initial set of markers.
+	err = writeMarkers(fd)
+	if err != nil {
+		return err
+	}
+
+	found, err := checkMarkers(fd)
+	if err != nil {
+		return err
+	}
+
+	if found != 3 {
+		return fmt.Errorf("Some of our initial markers weren't written properly")
+	}
+
+	// Start clearing the block.
+	_ = fd.Close()
+
+	// Attempt a secure discard run.
+	_, err = shared.RunCommand("blkdiscard", "--force", "--offset", strconv.FormatInt(blockOffset, 10), "--secure", blockPath)
+	if err == nil {
+		// Check if the markers are gone.
+		fd, err := os.Open(blockPath)
+		if err != nil {
+			return err
+		}
+
+		defer fd.Close()
+
+		found, err = checkMarkers(fd)
+		if err != nil {
+			return err
+		}
+
+		if found == 0 {
+			// All markers are gone, secure discard succeeded.
+			return nil
+		}
+
+		// Some markers were found, go to the next clearing option.
+		_ = fd.Close()
+	}
```

Comment:
That boilerplate code around `checkMarkers` is repeated 3 times. Any way to avoid the repetition?

Reply:
i tried a couple of ways of removing this repetition but it didnt result in much de-dupe in the end, and in one case it make the code less efficient by requiring another file handle. So I'm OK to leave as is.